### DISCUSSION TITLE
CSS modules - converting class names to camelCase automatically

### DISCUSSION THREAD

**Discussion by controversial** [vercel/next.js#11267] [Help]
Is it possible to configure the new built-in CSS modules support to automatically convert `.kebab-case` class names to `camelCase` when importing? `css-loader` supports this through the [`localsConvention` option](https://github.com/webpack-contrib/css-loader#localsconvention), but I'm not sure if/how it's possible to configure specific configuration options for `css-loader`.

**Comment by controversial** [comment]
I've come up with the following `next.config.js`, which works, but seems very brittle and very likely to break in a future Next.js release. Is there a better way than this?
```js
// next.config.js
module.exports = {
  webpack: (config) => {
    const { rules } = config.module;
    const styleRules = (rules.find((m) => m.oneOf && m.oneOf.find(({ test: reg }) => reg.test('file.css'))) || {}).oneOf;
    if (!styleRules) return config;
    const cssModuleRules = [
      styleRules.find(({ test: reg, use }) => reg.test('file.module.css') && use.loader !== 'error-loader'),
      styleRules.find(({ test: reg, use }) => reg.test('file.module.scss') && use.loader !== 'error-loader'),
    ].filter((n) => n);
    cssModuleRules.forEach((cmr) => {
      const cssLoaderConfig = cmr.use.find(({ loader }) => loader.includes('css-loader'));
      if (cssLoaderConfig && cssLoaderConfig.options) {
        cssLoaderConfig.options.localsConvention = 'camelCase';
      }
    });
    return config;
  },
};
```

With comments:
```js
// next.config.js
module.exports = {
  webpack: (config) => {
    const { rules } = config.module;
    // Find the array of "style rules" in the webpack config.
    // This is the array of webpack rules that:
    // - is inside a 'oneOf' block
    // - contains a rule that matches 'file.css'
    const styleRules = (rules.find((m) => m.oneOf && m.oneOf.find(({ test: reg }) => reg.test('file.css'))) || {}).oneOf;
    if (!styleRules) return config;
    // Find all the webpack rules that handle CSS modules
    // Look for rules that match '.module.css' and '.module.scss' but aren't being used to generate
    // error messages.
    const cssModuleRules = [
      styleRules.find(({ test: reg, use }) => reg.test('file.module.css') && use.loader !== 'error-loader'),
      styleRules.find(({ test: reg, use }) => reg.test('file.module.scss') && use.loader !== 'error-loader'),
    ].filter((n) => n); // remove 'undefined' values
    // Add the 'localsConvention' config option to the CSS loader config in each of these rules.
    cssModuleRules.forEach((cmr) => {
      // Find the item inside the 'use' list that defines css-loader
      const cssLoaderConfig = cmr.use.find(({ loader }) => loader.includes('css-loader'));
      if (cssLoaderConfig && cssLoaderConfig.options) {
        // Patch it with the new config
        cssLoaderConfig.options.localsConvention = 'camelCase';
      }
    });
    return config;
  },
};
```

**Reply by PaulTondeur** [reply]
Did you upgrade the css-loader? Cause for me it seems to run on version 4.3.0 which complains about localsConvention not matching the API schema (while running latest nextjs, 9.5.5)

Seems like a lot of hassle for something so relatively straightforward...

Curious to know if and how you upgraded the css-loader.

**Reply by timtyrrell** [reply]
The API changed a bit with the css-loader upgrade, so it is `exportLocalsConvention` now instead of `localsConvention` and one more level deeper, `options.modules` instead of just in `options`. This is the hack I am using combing about three of the other hacks 😅 and it works with 9.5.5 so far that I can tell

``` 
  const rules = config.module.rules
      .find((rule) => typeof rule.oneOf === "object")
      .oneOf.filter((rule) => Array.isArray(rule.use));

    rules.forEach((rule) => {
      rule.use.forEach((moduleLoader) => {
        if (
          moduleLoader.loader.includes("css-loader/dist") &&
          typeof moduleLoader.options.modules === "object"
        ) {
          moduleLoader.options = {
            ...moduleLoader.options,
            modules: {
              ...moduleLoader.options.modules,
              exportLocalsConvention: "camelCase", // https://github.com/webpack-contrib/css-loader#exportlocalsconvention
            },
          };
        }
      });
    });

    return config;
```

**Reply by PaulTondeur** [reply]
Thanks for your prompt reply Tim, it does work! Its a bit of a pity all this code is needed as a workaround for this moment, but glad you've put the work into it to sort it out ;-)

Hope the nextjs team will make this easier for us soon.

**Reply by SeinopSys** [reply]
To make this code work on Windows you will need to do this:

```diff
+const path = require('path');
 module.exports = {
     webpack: (config) => {
```
```diff
         rules.forEach((rule) => {
             rule.use.forEach((moduleLoader) => {
                 if (
-                    moduleLoader.loader.includes('css-loader/dist')
+                    moduleLoader.loader.includes(path.join('css-loader', 'dist'))
           && typeof moduleLoader.options.modules === 'object'
                 ) {
                     moduleLoader.options = {

**Reply by SeinopSys** [reply]
Bumping this to add that the code snippet needs to be updated to search for `css-loader/cjs` if it seems to have no effect, I upgraded a bunch of dependencies just now and it stopped working until I did that. I found what was actually being used by logging out `moduleLoader.loader` inside the `forEach` and watching the console output for anything with `css-loader` in it.

**Comment by lnikell** [comment]
I guess it should be enabled by default since with it you can use any case you want actually to access your styles. 
From js you can use styles.logoContainer however in your css it could be:
```
.logo-container
.logo_container
.logoContainer
```
and it will work.

**Reply by Defite** [reply]
By default, it is not working.

**Reply by IvanovRoman** [reply]
If writing `.logo_container` not work `styles.logoContainer` in `.tsx` files.

**Comment by wyattEpp** [comment]
Would love to see a less brittle approach for this as well.

**Comment by sidthesloth92** [comment]
Any updates on this one? webpack's css-loader provides a default option for the same. Would like to have this one implemented..😕

**Comment by Joozty** [comment]
Hi guys, can this be implemented please?

**Comment by tyteen4a03** [comment]
Bump - very useful feature that's sorely needed.

**Comment by murshex** [comment]
I want this... seriously not done yet?

**Comment by suhanw** [comment]
Bump!

**Comment by Kilerd** [comment]
I also faced this issue today, I typed `account-item` in css file, and then used it in component via `accountItem`, but it did not work for me.

**Comment by yura3d** [comment]
Follow my PR #22313, it brings the ability to configure CSS loader options, including set camel case for exported classes from CSS modules.

**Reply by RichardTMiles** [reply]
Reading the notes on this one made me sad... +1 trying to see support

**Reply by RichardTMiles** [reply]
https://webpack.js.org/loaders/css-loader/

Support is available using the `exportLocalsConvention: 'camelCase',` configuration override

**Comment by alejobh** [comment]
As March 31 with next update to `10.1` css module with sass webpack config apparently changed the way as `css-loader` comes.
I have a similar configuration as the comment above https://github.com/vercel/next.js/discussions/11267#discussioncomment-101154

The only change i made to make this work again is changing the includes `css-loader/dist` to `css-loader` and it worked again. 
Hope this helps anyone

**Reply by adampoczatek** [reply]
This was super helpful, thanks! For those who use BEM naming convention consider changing `exportLocalsConvention: 'camelCase'` to `exportLocalsConvention: 'dashes'` so that you do something like this:

my-component.module.scss:

```scss
.my-component { ... }

.my-component__child-element { ... }
```

JSX:

```tsx
import styles from './my-component.module.scss'

...

<div className={styles.myComponent}>
  <div className={styles.myComponent__childElement}>...</div>
</div>
```

**Comment by jzingh98** [comment]
Would really appreciate a fix to this problem. CSS is already toxic as it is to work with, and discovering this issue has made it all the more so

**Comment by zoeleu** [comment]
Holy crap, why is this not implemented yet??? Currently suffering with either not doing proper naming on SCSS or not doing proper naming on TypeScript.

**Comment by Melvin-Abraham** [comment]
The comments https://github.com/vercel/next.js/discussions/11267#discussioncomment-1758 and https://github.com/vercel/next.js/discussions/11267#discussioncomment-101154 _(along with changes specified in https://github.com/vercel/next.js/discussions/11267#discussioncomment-138721)_ doesn't seem to work anymore in Next.js 12. Is there any other recommended way or workaround to get this working?

**Reply by SeinopSys** [reply]
I use this method in my project, it's working well for me. Generally all I needed to do for the past 3-4 major versions was tweak that regex to make sure it matches the loader

https://github.com/SeinopSys/Rummage/blob/8fd919aa66b33ac42033692207a6d2fa8f23d6a8/utils/next-css-modules.js

**Reply by Melvin-Abraham** [reply]
@SeinopSys It works like a charm. Thanks a lot for the snippet... On a minor note, it might not work right away for windows platforms until the regex is tweaked a little:

```diff
- if (/css-loader\/(?:cjs|dist|src)/.test(moduleLoader.loader)) {
+ if (/css-loader[/\\](?:cjs|dist|src)/.test(moduleLoader.loader)) {
```

The above modification is required since Windows uses `\` (backslashes) as path delimeter.

**Comment by FreedCapybara** [comment]
At least as nasty, but in fewer lines:
```js
const _ = require('lodash');

module.exports = {
  reactStrictMode: true,
  webpack: config => {

    // camel-case style names from css modules
    const cssRules = _.find(config.module.rules, (rule) => !!rule.oneOf).oneOf;
    const cssLoaderRules = _.filter(cssRules, (rule) => _.includes(JSON.stringify(rule.use), 'css-loader'));
    const loaders = _.filter(_.flatMap(cssLoaderRules, 'use'), (loader) => !!loader.options.modules);
    for (const loader of loaders) {
      loader.options.modules.exportLocalsConvention = 'camelCase';
    }

    return config;
  }
};
```
Working for me with 12.1.0 👍

**Reply by Yizhachok** [reply]
Simpler method without any dependency like lodash.

```js
module.exports = {
	webpack: (config) => {
		// camel-case style names from css modules
		config.module.rules
			.find(({oneOf}) => !!oneOf).oneOf
			.filter(({use}) => JSON.stringify(use)?.includes('css-loader'))
			.reduce((acc, {use}) => acc.concat(use), [])
			.forEach(({options}) => {
				if (options.modules) {
					options.modules.exportLocalsConvention = 'camelCase';
				}
			});

		return config;
	},
};
```

**Reply by Scaelz** [reply]
> 



> Simpler method without any dependency like lodash.
> 
> ```js
> module.exports = {
> 	webpack: (config) => {
> 		// camel-case style names from css modules
> 		config.module.rules
> 			.find(({oneOf}) => !!oneOf).oneOf
> 			.filter(({use}) => JSON.stringify(use)?.includes('css-loader'))
> 			.reduce((acc, {use}) => acc.concat(use), [])
> 			.forEach(({options}) => {
> 				if (options.modules) {
> 					options.modules.exportLocalsConvention = 'camelCase';
> 				}
> 			});
> 
> 		return config;
> 	},
> };
> ```

where should i put this?

**Reply by diegosanz** [reply]
In `next.config.js`. More info: https://nextjs.org/docs/api-reference/next.config.js/introduction

**Reply by thefill** [reply]
> Simpler method without any dependency like lodash.
> 
> ```js
> module.exports = {
> 	webpack: (config) => {
> 		// camel-case style names from css modules
> 		config.module.rules
> 			.find(({oneOf}) => !!oneOf).oneOf
> 			.filter(({use}) => JSON.stringify(use)?.includes('css-loader'))
> 			.reduce((acc, {use}) => acc.concat(use), [])
> 			.forEach(({options}) => {
> 				if (options.modules) {
> 					options.modules.exportLocalsConvention = 'camelCase';
> 				}
> 			});
> 
> 		return config;
> 	},
> };
> ```

Why this is not a part of official config? This works like a charm...

**Reply by RichardTMiles** [reply]
The above + scss support
https://github.com/rainx/react-app-rewire-typings-for-css-module

**Comment by webdev163** [comment]
Simpler and less expensive way: just add this **fix.js** file to the main directory:

```js
const fs = require('fs');

const file = './node_modules/next/dist/build/webpack/config/blocks/css/loaders/modules.js';
fs.readFile(file, 'utf8', (err, data) => {
  const formatted = data.replace('exportLocalsConvention: "asIs"', 'exportLocalsConvention: "camelCase"');
  fs.writeFile(file, formatted, 'utf8', (error) => {
    if (error) { return console.log(error); }
  });
});
```

also add this script to package.json (this script will run automatically after installing dependencies):

```"postinstall": "node fix"```

**Comment by zoeleu** [comment]
Simpler fix, just use camelCase in CSS.

Sure, it's not pretty, but it's prettier than whatever patches are going on here.

**Comment by RichardTMiles** [comment]
This is my final config #beautiful ;)

```tsx
/* config-overrides.js */
// @link https://github.com/timarney/react-app-rewired/
const getLogger = require('webpack-log');
const rewireTypingsForCssModule = require("react-app-rewire-typings-for-css-module");
const {
    override,
    addWebpackPlugin, // <- added this line here
} = require('customize-cra')
const log = getLogger({name: 'config-overrides.js'});
const MiniCssExtractPlugin = require('mini-css-extract-plugin') // <- added this line here

const {
    defaultGetLocalIdent
} = require("css-loader/dist/utils")

module.exports = override(
    addWebpackPlugin(new MiniCssExtractPlugin()), // <- added this line here
    function (config, _env) {

        log.info('webpack env', process.env.NODE_ENV)

        // what's going on here? @link https://stackoverflow.com/questions/73551420/removing-hash-from-react-css-modules
        // noinspection JSUnusedGlobalSymbols
        config = rewireTypingsForCssModule.factory({
            modules: {
                exportLocalsConvention: 'camelCase',
                mode: 'local',
                localIdentName: 'production' === process.env.NODE_ENV ? '[hash:base64:5]' : '[path][name]__[local]--[hash:base64:5]',
                exportGlobals: true,
                getLocalIdent: (context, localIdentName, localName, options) => {

                    if (false === context.resourcePath.endsWith('/react/src/variables/bootstrap.module.scss')) {

                        localName = defaultGetLocalIdent(context, localIdentName, localName, options)
                            .replace("[local]", localName)

                    }

                    return localName;

                }

            }
        })(config); // do not pass env as it will cause the production build will not be modified

        // log.info('webpack configuration post config-overrides.js', JSON.stringify(config))


        return config;

    }
);
```

**Reply by zoeleu** [reply]
This is so scuffed just to write .selector-here and access it as styles.selectorHere.

Not only this, but you're using CRA libraries on NextJS configs, this may break with any updates.

**Comment by tobi1220** [comment]
Did anyone find a less-hacky solution to this yet?

**Comment by miraage** [comment]
2025, still not supported. Insanity..
